File: 5.1C-B/app/src/main/java/com/rorycd/istream/data/UserRepository.java
package com.rorycd.istream.data;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Handler;
import android.os.Looper;

import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class UserRepository {

    private static UserRepository INSTANCE;
    private static final String USER_ID = "user_id";
    private final UserDao userDao;
    private final PlaylistDao playlistDao;
    private final SharedPreferences sharedPrefs;
    private final ExecutorService executor = Executors.newSingleThreadExecutor();

    public interface LoginCallback {
        void onResult(User user);
    }

    public interface SignupCallback {
        void onResult(int userId);
    }

    private UserRepository(Context context) {
        userDao = UserDatabase.getInstance(context).userDao();
        playlistDao = UserDatabase.getInstance(context).playlistDao();
        sharedPrefs = context.getSharedPreferences("com.rorycd.istream.data", Context.MODE_PRIVATE);
    }

    public static UserRepository getInstance(Context context) {
        if (INSTANCE == null) {
            INSTANCE = new UserRepository(context);
        }
        return INSTANCE;
    }

    public void registerUser(String username, String fullName, String password, SignupCallback callback) {
        User newUser = new User(username, fullName, password);
        // Run on a separate thread
        executor.execute(() -> {
            int userId = (int) userDao.registerUser(newUser);
            sharedPrefs.edit().putInt(USER_ID, userId).apply();

            // Execute the callback on the main thread
            new Handler(Looper.getMainLooper()).post(() -> callback.onResult(userId));
        });
    }

    public void login(String username, String password, LoginCallback callback) {
        // Run a separate thread
        executor.execute(() -> {
            User user = userDao.login(username, password);
            if (user != null) {
                sharedPrefs.edit().putInt(USER_ID, user.id).apply();
            }
            // Execute the callback on the main thread
            new Handler(Looper.getMainLooper()).post(() -> callback.onResult(user));
        });
    }

    public int getCurrentUserId() {
        return sharedPrefs.getInt(USER_ID, -1);
    }

    public void addUrlToPlaylist(String url) {
        PlaylistItem item = new PlaylistItem(getCurrentUserId(), url);

        // Run a separate thread
        executor.execute(() -> {
            playlistDao.addItem(item);
        });
    }

    public List<String> getPlaylist() {
        return playlistDao.getPlaylistForUser(getCurrentUserId());
    }

    public boolean isLoggedIn() {
        return getCurrentUserId() != -1;
    }

    public void logout() {
        sharedPrefs.edit().remove(USER_ID).apply();
    }
}

File: 6.1D/app/src/main/java/com/rorycd/learningassistant/data/UserRepository.kt
package com.rorycd.learningassistant.data

import android.content.Context
import android.content.SharedPreferences
import androidx.core.content.edit
import kotlinx.coroutines.flow.Flow

const val USER_ID_KEY = "current_user_id"
const val SHARED_PREFS_NAME = "prefs"

class UserRepository(private val userDao: UserDao, private val context: Context) {
    val sharedPrefs: SharedPreferences = context.getSharedPreferences(SHARED_PREFS_NAME, Context.MODE_PRIVATE)

    suspend fun login(username: String, password: String) : Boolean {
        // Check user exists
        val user = userDao.login(username, password)

        // Save user ID to shared prefs
        if (user != null)
            sharedPrefs.edit { putInt(USER_ID_KEY, user.id) }

        return user != null
    }

    fun logOut() {
        sharedPrefs.edit { putInt(USER_ID_KEY, -1) }
    }

    suspend fun register(
        username: String,
        password: String,
        email: String,
        imgUri: String?
    ): Boolean {
        val newUser = User(0, username, password, email, interests = null, imgUri = imgUri)
        try {
            userDao.insert(newUser)
            login(username, password)
            return true
        } catch(e: Exception) {
            return false
        }
    }

    fun currentUserId(): Int {
        return sharedPrefs.getInt(USER_ID_KEY, -1)
    }

    suspend fun getCurrentUser(): User? {
        val id = currentUserId()
        val user = userDao.getUserById(id)
        return user
    }

    fun getCurrentUserFlow(): Flow<User?> {
        return userDao.getUserFlowById(currentUserId())
    }

    suspend fun userExists(username: String): Boolean {
        val user = userDao.getUserByUsername(username)
        return user != null
    }

    fun isLoggedIn() : Boolean {
        return currentUserId() != -1
    }

    suspend fun addInterests(interests: Set<String>) {
        val interestList = interests.toList()
        userDao.updateInterests(currentUserId(), interestList)
    }
}
